### DISCUSSION TITLE
FastAPI Query with dataclass fails using alias

### DISCUSSION THREAD

**Discussion by fabriba** [fastapi/fastapi#8535] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, Depends, Query
from typing import Optional
from pydantic.dataclasses import dataclass


app = FastAPI()

@dataclass
class Catz:
    qqq: Optional[str] = Query(None, alias="q")

@app.get("/productz/")
def search_products(query: Catz = Depends(Catz)  ):
    products = [{"name": "Computer"}, {"name": "HDD"}]
    if not query.qqq:
        query.qqq = ""
    return {"query": query, "results": [product for product in products if query.qqq in product["name"]]}


@dataclass
class Cats:
    qqq: Optional[str] = Query(None )

@app.get("/products/")
def search_products(query: Cats = Depends(Cats)  ):
    products = [{"name": "Computer"}, {"name": "HDD"}]
    if not query.qqq:
        query.qqq = ""
    return {"query": query, "results": [product for product in products if query.qqq in product["name"]]}




if __name__ == "__main__":
    uvicorn.run("main:app", port=11978, log_level="info", reload=True)
```

### Description

when I use the service via curl, I get the following outputs:

the expected behaviour with the endpoint /products/ that has no aliases
>> curl -X 'GET'  'http://localhost:11978/products/?qqq=H'  -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":"H"},"results":[{"name":"HDD"}]}

not the expected behaviour with the endpoint /productz/ that has no aliases (regardless of me using a query parameter with its own name or with the alias I have in the code)
>> curl -X 'GET'  'http://localhost:11978/productz/?qqq=H'  -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":""},"results":[{"name":"Computer"},{"name":"HDD"}]}

>> curl -X 'GET'  'http://localhost:11978/productz/?q=H'  -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":""},"results":[{"name":"Computer"},{"name":"HDD"}]}

the problem goes away if I use 
from dataclasses import dataclass
but that has other drawbacks (especially for more complex and nested dataclasses) as stated in FastAPI documentation.

### Operating System

Linux, macOS

### FastAPI Version

0.75.2

### Python Version

3.10.4

**Comment by jonatasoli** [comment]
Hello,

You try this case changing dataclass to pydantic model? I try use pydantic BaseModel and found and using dataclass my output is diferent to you

>curl -X 'GET' 'http://localhost:11978/products/?qqq=H' -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":"H"},"results":[{"name":"HDD"}]}

-This different:
> curl -X 'GET' 'http://localhost:11978/productz/?qqq=H' -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":"H"},"results":[{"name":"HDD"}]}⏎

> curl -X 'GET' 'http://localhost:11978/productz/?q=H' -H 'accept: application/json' -H 'api-version: 1.0' ; echo
{"query":{"qqq":""},"results":[{"name":"Computer"},{"name":"HDD"}]}

**Comment by JarroVGIT** [comment]
I think this is a bug. The documentation clearly states that the `pydantic.dataclasses` is a drop-in replacement for `dataclasses`, which it isn't in this case. 

I had a hunch, and adding this to your code will allow you to use the `alias` to set the `qqq` property using `q` (but will not allow you to use `qqq` anymore!):

```python
from pydantic import BaseConfig
BaseConfig.allow_population_by_field_name = True
```

That is very strange behaviour. Normally, the above would allow you to use both `qqq` and `q` to instantiate a `Catz` object. However, when using in your code, this allows to use `q` but not `qqq`. Which is better than what you were experiencing (e.g. neither `q` nor `qqq` were respected) but still, very weird behaviour indeed.

**Comment by YuriiMotov** [comment]
Still doesn't work in fastapi 0.115.12:

Test in details:

<details>

```python
from dataclasses import dataclass

import pytest
from fastapi import Depends, FastAPI, Query
from fastapi.testclient import TestClient
from pydantic.dataclasses import dataclass as pydantic_dataclass

app = FastAPI()

# Regular dataclass

@dataclass
class Catz:
    qqq: str = Query(..., alias="q")

@app.get("/dataclass")
def _dataclass(query: Catz = Depends()  ):
    return {"query": query.qqq}


# Pydantic dataclass

@pydantic_dataclass
class CatzPy:
    qqq: str = Query(..., alias="q")


@app.get("/py-dataclass")
def _py_dataclass(query: CatzPy = Depends()  ):
    return {"query": query.qqq}


# Test

@pytest.mark.parametrize("path", ["/dataclass", "/py-dataclass"])
def test_(path: str):
    client = TestClient(app, raise_server_exceptions=False)
    resp = client.get(path, params={"q": "123"})
    assert resp.status_code == 200  # status code 500 for /py-dataclass
    assert resp.json() == {"query": "123"}

```

</details>

Error
```
>                       result = context.run(func, *args)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^
E                       pydantic_core._pydantic_core.ValidationError: 1 validation error for CatzPy
E                       qqq
E                         Field required [type=missing, input_value=ArgsKwargs((), {'q': '123'}), input_type=ArgsKwargs]
E                           For further information visit https://errors.pydantic.dev/2.11/v/missing
```

Workarounds:
* use `Field` instead of `Query`
```py
@pydantic_dataclass
class CatzPy:
    qqq: str = Field(..., alias="q")
```
* use `validation_alias` instead of `alias`
```py
@pydantic_dataclass
class CatzPy:
    qqq: str = Query(..., validation_alias="q")
```
